src/Host/Monitor/Impl/CredentialManager.cs

```diff
@@ -82,69 +67,87 @@ public class CredentialManager {
+                            creds.attributeCount = 0;
+                            creds.persist = (int)CredPersist.LOCAL_MACHINE;
+                            byte[] bpassword = Encoding.Unicode.GetBytes(password.ToString());
+                            creds.credentialBlobSize = bpassword.Length;
+                            creds.credentialBlob = Marshal.StringToCoTaskMemUni(password.ToString());
+
+                            if (!CredWrite(ref creds, 0)) {
+                                // Failed to save credentials. Do another attempt by showing user the CredUI with the errorCode set.
+                                errorCode = Marshal.GetLastWin32Error();
+                                if (token != IntPtr.Zero) {
+                                    CloseHandle(token);
                                 }
+                                continue;
                             }
-                        } finally {
+                        } else {
+                            // Failed to login user. Do another attempt by showing user the CredUI with the errorCode set.
+                            errorCode = Marshal.GetLastWin32Error();
                             if (token != IntPtr.Zero) {
                                 CloseHandle(token);
                             }
+                            continue;
                         }
 
+                        // Credentials received successfully cleanup.
+                        if (token != IntPtr.Zero) {
+                            CloseHandle(token);
+                        }
+
+                        // exit loop, another attempt is not needed.
                         break;
-                    }
+                    } else {
+                        // CredUIPromptForWindowsCredentials failed to load some component
+                        if (credentialBuffer != IntPtr.Zero) {
+                            Marshal.FreeCoTaskMem(credentialBuffer);
+                        }
 
-                    // CredUIPromptForWindowsCredentials failed to load some component
-                    if (ret != 0) {
                         throw new Win32Exception(string.Format(Resources.Error_CredUIFailedToLoad, ret));
                     }
                 }
-            });
-        }
 
-        public static Task RemoveCredentialsAsync() {
-            return Task.Run(() => {
-                CredDelete(BrokerUserCredName, CredType.GENERIC, 0);
+                if(attempts >= 3) {
+                    MainWindow.SetStatusText(Resources.Info_TooManyLoginAttempts);
+                }
             });
         }
 
+        public static void RemoveCredentials() {
+            CredDelete(BrokerUserCredName, CredType.GENERIC, 0);
+        }
 
-        public static Task SetCredentialsOnProcess(ProcessStartInfo psi) {
+        public static Task SetCredentialsOnProcessAsync(ProcessStartInfo psi) {
             return Task.Run(()=> {
```

Comment:
Is the purpose of `Task.Run` to force the code to run on background thread and drop any context, even if the caller is on UI thread? If so, `await TaskUtilities.SwitchToBackgroundThread()` is the preferred way to do so.